crates/history/src/any.rs

```diff
@@ -13,8 +15,10 @@ use crate::{error::HistoryResult, query::ToQuery};
 #[derive(Clone, PartialEq, Debug)]
 pub enum AnyHistory {
     /// A Browser History.
+    #[cfg(not(target_os = "wasi"))]
```

Comment:
What are the effects of enabling BrowserHistory and MemoryHistory for wasi?

I think wasi should have BrowserHistory and MemoryHistory enabled like other targets.
This helps to avoid additional feature flags for downstream crates and will lint the code like other targets.

Reply:
> What are the effects of enabling BrowserHistory and MemoryHistory for wasi?
> 
> I think wasi should have BrowserHistory and MemoryHistory enabled like other targets. This helps to avoid additional feature flags for downstream crates and will lint the code like other targets.

If don't, `BrowserHistory` is directly dependent on `wasm-bindgen`. It will cause additional symbols like `__wbindgen_placeholder__::xxx` to appear in the compiled result WASM, which is invalid in WASI.

These changes that I've made to some other repositories recently were also made to fix this.

Reply:
I think WASI has been disabled in https://github.com/rustwasm/wasm-bindgen/pull/3233?

Reply:
> I think WASI has been disabled in [rustwasm/wasm-bindgen#3233](https://github.com/rustwasm/wasm-bindgen/pull/3233)?

It seems like it will still try to import symbols if you actively reference it. This is a bit different from the situation where `wasm-bindgen` passively imports utility functions related to heap allocation.

Personally, I feel that it is not a bad thing to proactively limit the compilation targets available to the code. Because this allows the compiler to warn developers faster.

Reply:
> It seems like it will still try to import symbols if you actively reference it.

I think you may need to run `cargo update`. Since this is only released in 0.2.88 last month.

I tried the following code:

```rust
fn main() {
    js_sys::Date::now();
}
```

<img width="916" alt="image" src="https://github.com/rustwasm/gloo/assets/11693215/3bc18d59-6541-498c-82e1-b81d74c04ab8">


> Personally, I feel that it is not a bad thing to proactively limit the compilation targets available to the code. Because this allows the compiler to warn developers faster.

This is a design decision made all the way up to wasm-bindgen.
One of the reasons is that you can write code targeting both `wasm` (client) and native targets (server) at the same time without having to reconfigure toolchain.

<img width="775" alt="image" src="https://github.com/rustwasm/gloo/assets/11693215/2eae64b5-5ce4-4cf2-9aa1-d183ae91fdc6">

Reply:
> > It seems like it will still try to import symbols if you actively reference it.
> 
> I think you may need to run `cargo update`. Since this is only released in 0.2.88 last month.
> 
> I tried the following code:
> 
> ```rust
> fn main() {
>     js_sys::Date::now();
> }
> ```
> 
> <img alt="image" width="916" src="https://private-user-images.githubusercontent.com/11693215/289228557-3bc18d59-6541-498c-82e1-b81d74c04ab8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwOTA3MjksIm5iZiI6MTcwMjA5MDQyOSwicGF0aCI6Ii8xMTY5MzIxNS8yODkyMjg1NTctM2JjMThkNTktNjU0MS00OThjLTgyZTEtYjgxZDc0YzA0YWI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA5VDAyNTM0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhZTFiNDI1NDg5MjBhZWJjYjg1YmQ4MWJjY2RkZTEwZTYxMjU3NWVjOTQ5ZmEyZmY3OWI4ZGQxN2NjOGJkOTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bN9OAJ4ALNFvkSsCJz9843F87ke2RBXI0SMgEORFJ9Y">
> > Personally, I feel that it is not a bad thing to proactively limit the compilation targets available to the code. Because this allows the compiler to warn developers faster.
> 
> This is a design decision made all the way up to wasm-bindgen. One of the reasons is that you can write code targeting both `wasm` (client) and native targets (server) at the same time without having to reconfigure toolchain.
> 
> <img alt="image" width="775" src="https://private-user-images.githubusercontent.com/11693215/289228816-2eae64b5-5ce4-4cf2-9aa1-d183ae91fdc6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwOTA3MjksIm5iZiI6MTcwMjA5MDQyOSwicGF0aCI6Ii8xMTY5MzIxNS8yODkyMjg4MTYtMmVhZTY0YjUtNWNlNC00Y2YyLTlhYTEtZDE4M2FlOTFmZGM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA5VDAyNTM0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVhOTc1MGY4MGFmZWYxMjU5YmQ1NGY4NDVkOTU0YjgwNDk0MzVjMDUwNTk0OGI4MTcxZDlkNjA4ODgzZmVkYmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.s5y5xKI8ZX7cRkc4BoVpQ7Buonsvo54U3ahDrDjJjTA">

Oh……

But this update requires updating the CI of all projects compiled using `wasm-bindgen`. I'll still keep [the command line arguments](https://github.com/bytecodealliance/wasmtime/pull/4312) `--trap-unknown-imports` added to the commit's part about calling `wasmtime` CLI.

Reply:
Updated on https://github.com/yewstack/yew/pull/3534/commits/66996868523716ef3743ccd24e880c3ba27d17e3

Reply:
But this update requires updating the CI cache of all projects compiled using wasm-bindgen. 

We can restraint `wasm-bindgen = “0.2.88"` for gloo-history, all downstream crates will automatically update to use 0.2.88 or later once they use a gloo-history version that supports wasi.

Reply:
> > But this update requires updating the CI cache of all projects compiled using wasm-bindgen.
> 
> We can restraint `wasm-bindgen = "0.2.88"` for gloo-history, all downstream crates will automatically update to use 0.2.88 or later once they use a gloo-history version that supports wasi.

I've create a PR #423 to fix it.

Reply:
@langyo Since 0.2.88 solves the WASI issues, would it be possible to create a pull request to remove the feature flags from gloo-history so we do not subsequently introduce more feature flags into Yew?

Reply:
> @langyo Since 0.2.88 solves the WASI issues, would it be possible to create a pull request to remove the feature flags from gloo-history so we do not subsequently introduce more feature flags into Yew?

I think it's better not to remove it yet. Keeping some flags will help users who compile with WASI to quickly locate the cause of the error.

But I will actively consider this matter... Perhaps there's no need to resort to strong measures to circumvent this problem.

Reply:
> Keeping some flags will help users who compile with WASI to quickly locate the cause of the error.

I think it is important to ensure consistency between gloo and wasm-bindgen handling.

When users wrongly uses gloo-history under wasi with client side rendering code, they pretty much uses wasm-bindgen and web-sys wrongly as well. So the user will inevitably be facing runtime errors even if gloo-history uses compiler flags.

I have tried to apply a similar approach with target flags to make Yew `Send`, and it resulted in a very large number of feature flags being used and it will require every user who writes SSR to apply a large number of feature flags.
I have reached a conclusion that using feature flags as unfeasible for end users.

In addition, it will eventually be possible for wasm32-wasi to run under browser environments.
These feature flags will need to be removed to provide that support, which will be another breaking changes.

Reply:
Your opinion is also very reasonable. I thought about it for a while and I'll try to roll back a part of the code later.


On the possibility of WASI running in the browser. Personally, I don't think it would be appropriate for WASI to be a well-designed virtual machine interface that would be arbitrarily introduced as a non-standard interface. It's not for technical reasons, but because of some historical lessons... If some non-standard interfaces are adopted on a large scale, it will be very troublesome to standardize for them in the future.

As a recent example, `wasmer` has actually made an interface that is slightly different from standard WASI. Even though this made it become the first WASM runtime written in Rust to be able to communicate over the network, it came at the cost of introducing a large number of specialized patch libraries, even including the Rust compiler.

Reply:
> On the possibility of WASI running in the browser.

Browsers are already on track to support WASI.
What is not yet decided is that whether and how wasm-bindgen wants to support wasi.

https://bugzilla.mozilla.org/show_bug.cgi?id=1701197